src/ui/menu.c

```diff
@@ -489,8 +456,43 @@ meta_window_menu_new   (MetaFrames         *frames,
   return menu;
 }
 
+static GdkEvent *
+menu_popup_event (gint root_x, gint root_y, guint button, guint32 timestamp,
+                  GdkWindow *window)
+{
+    GdkEvent  *event;
+    GdkSeat   *seat;
+    GdkDevice *device;
+
+    event = gtk_get_current_event ();
+
+    if (event != NULL)
+        event = gdk_event_copy (event);
+    else
+    {
+        /* Create fake event since menu can be show without any events */
+
+        seat = gdk_display_get_default_seat (gdk_window_get_display (window));
+        device = gdk_seat_get_pointer (seat);
+
+        event = gdk_event_new (GDK_BUTTON_PRESS);
+        event->button.window = g_object_ref (window);
+        event->button.time = timestamp;
+        event->button.x = event->button.x_root = root_x;
+        event->button.y = event->button.y_root = root_y;
+        event->button.button = button;
+        event->button.device = device;
+        gdk_event_set_device (event, device);
```

Comment:
AFAICT, all (internal) code paths leading to a `meta_window_menu_popup()` start either with a `GdkEvent` or an `XEvent`.  If now a `GdkEvent` is necessary, couldn't we just forward taht to here (actually making the API itself simpler), and provide a `gdk_event_from_x_event()` kind of function that would translate one to the other? (this probably exists somewhere in GDK itself, so that could be copied over if need be)

Reply:
Doing so will introduce a lot of code and increase the scope of changes，The implementation of ```gdk_event_from_x_event()``` is similar to that of ```menu_popup_event```.We can use existing parameters to build event，This has little impact on the whole project